### PR TITLE
Spyglass deduplicates same test, and supports displaying flaky tests

### PR DESCRIPTION
**Pull request by chaodaiG**
spyglass displays test result multiple times when a test ran multiple times in a Prow job, see example here: https://prow.knative.dev/view/gcs/knative-prow/pr-logs/pull/knative_test-infra/1615/pull-knative-test-infra-unit-tests/1215399931288752132

The above use case is trying to discover test flakiness by `go test -count=5`, and the desired behavior is a deduplicated result in spyglass:
- If a test succeeds 5 times then only 1 of them is shown
- Same as failed and skipped
- If a flake is discovered, it would be better classified as flaky instead of some in failed and some in succeeded

This PR addresses the above problem by:
- Deduplicate test cases that appeared more than once in the SAME junit result file
- Add a flaky category aside from succeed, failed, and skipped
- Also added flaky category in UI

By the way did some clean ups:
- Break up junit lens a little bit to make it easier testable
- Add unit test for junit lens (there was none before)

/assign @Katharine @fejta

### CODE REVIEW

File: prow/spyglass/lenses/dumplenses.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+package lenses
+
+import (
+	"bytes"
+	"context"
+	"io/ioutil"
+)
+
+type FakeArtifact struct {
```
Comment (fejta) [inline prow/spyglass/lenses/dumplenses.go]: Please put this back into a test file. This should not ship with the code.

Comment (coryrc) [reply prow/spyglass/lenses/dumplenses.go]: The problem is it can't then be used from the `junit` subfolder.
The solution is https://stackoverflow.com/a/38715325

Comment (fejta) [reply prow/spyglass/lenses/dumplenses.go]: Copy the code in both locations or put it in its own testing package.

This doesn't belong in prod

Comment (chaodaiG) [reply prow/spyglass/lenses/dumplenses.go]: sg, I'll copy the code

Comment (chaodaiG) [reply prow/spyglass/lenses/dumplenses.go]: done

**Comment by chaodaiG** [comment]
It's been 25 minutes since I pushed and Prow jobs still haven't started

**Comment by fejta** [comment]
https://kubernetes.slack.com/archives/C7J9RP96G/p1585339001014200

**Comment by chaodaiG** [comment]
> https://kubernetes.slack.com/archives/C7J9RP96G/p1585339001014200

Have joined in k8s slack yet, let me try

**Comment by chaodaiG** [comment]
> > https://kubernetes.slack.com/archives/C7J9RP96G/p1585339001014200
> 
> Have joined in k8s slack yet, let me try

Just joined. Thanks for the info

File: prow/spyglass/lenses/junit/junit.css [hunk 2]
```diff
@@ -25,7 +25,7 @@
   text-align: right;
 }
 
-td.failed {
+td.failed, td.flaky {
```
Comment (fejta) [inline prow/spyglass/lenses/junit/junit.css]: flaky should be purple, not red

File: prow/spyglass/lenses/junit/lens.go [hunk 3]
```diff
@@ -136,48 +178,78 @@ func (lens Lens) Body(artifacts []lenses.Artifact, resourceDir string, data stri
 	}
 	sort.Slice(results, func(i, j int) bool { return results[i].path < results[j].path })
 
-	jvd := struct {
-		NumTests int
-		Passed   []TestResult
-		Failed   []TestResult
-		Skipped  []TestResult
-	}{}
+	jvd := JVD{}
```
Comment (fejta) [inline prow/spyglass/lenses/junit/lens.go]: Prefer `var jvd JVD` unless you are initializing something

https://dave.cheney.net/practical-go/presentations/qcon-china.html#_use_a_consistent_declaration_style

File: prow/spyglass/lenses/junit/lens.go [hunk 4]
```diff
@@ -120,13 +148,27 @@ func (lens Lens) Body(artifacts []lenses.Artifact, resourceDir string, data stri
 				for _, subSuite := range suite.Suites {
 					record(subSuite)
 				}
+
 				for _, test := range suite.Results {
-					result.junit = append(result.junit, test)
+					test := test
```
Comment (fejta) [inline prow/spyglass/lenses/junit/lens.go]: redundant

File: prow/spyglass/lenses/junit/lens.go [hunk 5]
```diff
@@ -136,48 +178,78 @@ func (lens Lens) Body(artifacts []lenses.Artifact, resourceDir string, data stri
...
+	jvd := JVD{}
+
 	for _, result := range results {
 		if result.err != nil {
 			continue
 		}
-		for _, test := range result.junit {
-			if test.Failure != nil {
+		for _, tests := range result.junit {
+			var skipped, passed, failed, flaky bool
```
Comment (fejta) [inline prow/spyglass/lenses/junit/lens.go]: I'd prefer if these are on their own line.

Consider using 

```golang
var (
  skipped bool
  cumFailure []string
)
```

File: prow/spyglass/lenses/junit/lens.go [hunk 6]
```diff
@@ -136,48 +178,78 @@ func (lens Lens) Body(artifacts []lenses.Artifact, resourceDir string, data stri
...
+					}
+					if !flaky {
+						failed = true
+					}
+					cumFailure = append(cumFailure, *test.Failure)
+					// Use the last failed result for flaky test, or same test
+					// failed multiple times
+					combinedTest = test
+				} else {
+					if failed {
```
Comment (fejta) [inline prow/spyglass/lenses/junit/lens.go]: combine with previous line

File: prow/spyglass/lenses/junit/lens.go [hunk 7]
```diff
@@ -88,15 +97,34 @@ type TestResult struct {
...
+
+	var buf bytes.Buffer
+	if err := junitTemplate.ExecuteTemplate(&buf, "body", jvd); err != nil {
+		logrus.WithError(err).Error("Error executing template.")
+	}
+
+	return buf.String()
+}
+
+func (lens Lens) getJvd(artifacts []lenses.Artifact) JVD {
```
Comment (fejta) [inline prow/spyglass/lenses/junit/lens.go]: We may want to allow people to toggle between the current and previous functionality via a flag.

File: prow/spyglass/lenses/junit/lens.go [hunk 8]
```diff
@@ -136,48 +178,78 @@ func (lens Lens) Body(artifacts []lenses.Artifact, resourceDir string, data stri
...
+						passed = true
+						// Don't use a succeeded test result for flaky test
+						combinedTest = test
+					}
+				}
+			}
+
+			if len(cumFailure) > 0 {
+				failure := strings.Join(cumFailure,
+					"\n\n---Separation line for tests that failed reruns---\n\n")
```
Comment (fejta) [inline prow/spyglass/lenses/junit/lens.go]: This seems pretty hacky. IMO each should preserve the separate junit.Result like it does currently. It may be necessary/worthwhile to allow people to expand each flaky set of tests to the individual results.

File: prow/spyglass/lenses/junit/lens.go [hunk 9]
```diff
@@ -120,13 +148,27 @@ func (lens Lens) Body(artifacts []lenses.Artifact, resourceDir string, data stri
...
+
 				for _, test := range suite.Results {
-					result.junit = append(result.junit, test)
+					test := test
+					// There are cases where multiple entries of exactly the same
+					// testcase in a single junit result file, this could result
+					// from reruns of test cases by `go test --count=N` where N>1.
+					// Deduplicate them here in this case, and classify a test as being
+					// flaky if it both succeeded and failed
+					testFullName := fmt.Sprintf("%s...%s...%s", suite.Name, test.ClassName, test.Name)
```
Comment (fejta) [inline prow/spyglass/lenses/junit/lens.go]:
```golang
type foo struct {
  file string
  class string
  test string
}
```

Seems like a more appropriate key than string mangling.

File: prow/spyglass/lenses/junit/lens.go [hunk 10]
```diff
@@ -120,13 +148,27 @@ func (lens Lens) Body(artifacts []lenses.Artifact, resourceDir string, data stri
...
+					if _, ok := groups[testFullName]; !ok {
+						groups[testFullName] = make([]junit.Result, 0)
```
Comment (fejta) [inline prow/spyglass/lenses/junit/lens.go]: This shouldn't be necessary

```golang
var foo []string = nil
foo = append(foo, "hello")
```
works just fine

File: prow/spyglass/lenses/junit/lens.go (hunk 7 above)
Comment (fejta) [reply prow/spyglass/lenses/junit/lens.go]: Sounds like it might be fine to just ship it: https://kubernetes.slack.com/archives/CDECRSC5U/p1585349532027400?thread_ts=1585348940.027100&cid=CDECRSC5U

Comment (chaodaiG) [reply prow/spyglass/lenses/junit/lens.go]: Great to know. Will update based on other comments

**Comment by stevekuznetsov** [comment]
Any screenshots or PR-based webpages to poke around it?

File: prow/spyglass/lenses/junit/junit.css (hunk 2 above)
Comment (chaodaiG) [reply prow/spyglass/lenses/junit/junit.css]: done

File: prow/spyglass/lenses/junit/lens.go (hunk 4 above)
Comment (chaodaiG) [reply prow/spyglass/lenses/junit/lens.go]: that's right, done

File: prow/spyglass/lenses/junit/lens.go (hunk 10 above)
Comment (chaodaiG) [reply prow/spyglass/lenses/junit/lens.go]: interesting, seems like I've had the wrong impression for a while. done and thanks!

File: prow/spyglass/lenses/junit/lens.go (hunk 9 above)
Comment (chaodaiG) [reply prow/spyglass/lenses/junit/lens.go]: great to learn. done

File: prow/spyglass/lenses/junit/lens.go (hunk 3 above)
Comment (chaodaiG) [reply prow/spyglass/lenses/junit/lens.go]: done

File: prow/spyglass/lenses/junit/lens.go (hunk 5 above)
Comment (chaodaiG) [reply prow/spyglass/lenses/junit/lens.go]: done

File: prow/spyglass/lenses/junit/lens.go (hunk 6 above)
Comment (chaodaiG) [reply prow/spyglass/lenses/junit/lens.go]: done

File: prow/spyglass/lenses/junit/lens.go (hunk 8 above)
Comment (chaodaiG) [reply prow/spyglass/lenses/junit/lens.go]: Totally agree that it's hacky, and it's certainly a good idea to allow people expand each flaky set of tests, since this causes no regression at all. I will think about how to make this work

Comment (fejta) [reply prow/spyglass/lenses/junit/lens.go]: Katharine and michelle192837 should be able to point you in the right direction for the necessary typescript changes.

Comment (chaodaiG) [reply prow/spyglass/lenses/junit/lens.go]: Got plenty of help from them, and I think it's working now. This is almost to the best of my html knowledge

**Comment by chaodaiG** [comment]
> Any screenshots or PR-based webpages to poke around it?

Example screenshot:
Screen Shot 2020-04-01 at 8 08 15 AM

Expanding flaky section:
Screen Shot 2020-04-01 at 8 10 52 AM

Expanding each occurance of a single flaky test:
Screen Shot 2020-04-01 at 8 11 04 AM

Succeeded is not affected:
Screen Shot 2020-04-01 at 8 10 04 AM

Failed is similar to flaky, allow expanding as well:
Screen Shot 2020-04-01 at 8 11 23 AM

File: prow/spyglass/lenses/junit/template.html [hunk 11]
```diff
@@ -21,38 +22,127 @@
...
-          <tr class="hidden failure-text">
-            <td colspan="2" class="mdl-data-table__cell--non-numeric">
-              <div>{{$test.Junit.Failure}}</div>
-              {{if $test.Junit.Output}}
-              <a href="#" class="open-stdout">open stdout<i class="material-icons" style="font-size: 1em; vertical-align: middle; padding-left: 3px;">open_in_new</i></a>
-              <pre style="display: none;">{{$test.Junit.Output}}</pre>
-              {{end}}
+          <tr class="hidden">
+            <td>
+              <table  class="flaky-layout">
```
Comment (michelle192837) [inline prow/spyglass/lenses/junit/template.html:104]: nit: Think there's an extra space?

File: prow/spyglass/lenses/junit/lens.go [hunk 12]
```diff
@@ -136,48 +193,65 @@ func (lens Lens) Body(artifacts []lenses.Artifact, resourceDir string, data stri
...
-	}{}
+	var jvd JVD
+
 	for _, result := range results {
 		if result.err != nil {
 			continue
 		}
-		for _, test := range result.junit {
-			if test.Failure != nil {
+		for _, tests := range result.junit {
```
Comment (michelle192837) [inline prow/spyglass/lenses/junit/lens.go:202]: The logic for figuring out flakiness seems complicated enough to get its own function (which should make testing it a bit easier too?)

File: prow/spyglass/lenses/junit/lens.go [hunk 13]
```diff
@@ -136,48 +193,65 @@ func (lens Lens) Body(artifacts []lenses.Artifact, resourceDir string, data stri
 	}
 	sort.Slice(results, func(i, j int) bool { return results[i].path < results[j].path })
 
-	jvd := struct {
-		NumTests int
-		Passed   []TestResult
-		Failed   []TestResult
-		Skipped  []TestResult
-	}{}
+	var jvd JVD
```
[the remark this answers is not in the document]
Comment (chaodaiG) [reply prow/spyglass/lenses/junit/lens.go:196]: To be honest I had no idea, my initial guess was `Junit Vsomething Data`

**Comment by chaodaiG** [comment]
Will leave some time for other reviewers, if no other concerns I'll merge it first thing tomorrow morning

File: prow/spyglass/lenses/junit/template.html [hunk 14]
```diff
@@ -21,38 +22,127 @@
...
+                  <a href="#" class="open-stdout">open stdout<i class="material-icons" style="font-size: 1em; vertical-align: middle; padding-left: 3px;">open_in_new</i></a>
+                  <pre style="display: none;">{{$firstTest.Output}}</pre>
+                  {{end}}
+                </td>
+              </tr>
+            </table>
+          </td>
+        </tr>
+      {{else}}
+      <tr>
```
Comment (michelle192837) [inline prow/spyglass/lenses/junit/template.html:48]: nit: This block should be at the same indentation level as in the 'if' above. (same with other blocks under if/else controls, for consistency).

File: prow/spyglass/lenses/junit/template.html [hunk 15]
```diff
@@ -21,38 +22,127 @@
     </tr>
     <tbody id="failed-tbody">
     {{range $ix, $test := .Failed}}
+      {{$numTest := len $test.Junit}}
+      {{$firstTest := index $test.Junit 0}}
+      {{if eq $numTest 1}}
```
Comment (michelle192837) [inline prow/spyglass/lenses/junit/template.html:27]: Would be great to abstract some of the common HTML out of the if/else block (if possible, it looks like there's some common lines between them but I may be missing something that means they can't be factored out)

File: prow/spyglass/lenses/junit/junit.css [hunk 16]
```diff
@@ -26,7 +26,11 @@
 }
 
 td.failed {
-  color: #ff4040;
+  color: #ff0000;
```
[the remark this answers is not in the document]
Comment (chaodaiG) [reply prow/spyglass/lenses/junit/junit.css]: no it doesn't, fixed

File: prow/spyglass/lenses/junit/junit.css [hunk 17]
```diff
@@ -26,7 +26,11 @@
...
+}
+
+td.flaky {
+  color: #800080;
```
[the remark this answers is not in the document]
Comment (chaodaiG) [reply prow/spyglass/lenses/junit/junit.css]: lgtm, changed

File: prow/spyglass/lenses/junit/lens.go (hunk 12 above)
Comment (chaodaiG) [reply prow/spyglass/lenses/junit/lens.go:202]: would prefer not, transmitting these 4 states between 2 functions seems a bit too much to me

File: prow/spyglass/lenses/junit/template.html (hunk 14 above)
Comment (chaodaiG) [reply prow/spyglass/lenses/junit/template.html:48]: done with indentation

File: prow/spyglass/lenses/junit/template.html (hunk 11 above)
Comment (chaodaiG) [reply prow/spyglass/lenses/junit/template.html]: nice catch, fixed

**Comment by chaodaiG** [comment]
@michelle192837 , you are definitely right on the colors, they look much better now:
Screen Shot 2020-04-02 at 4 35 26 PM

File: prow/spyglass/lenses/junit/template.html (hunk 15 above)
Comment (chaodaiG) [reply prow/spyglass/lenses/junit/template.html:27]: I just checked, it didn't seem to me that there is much could be refactored out
